pkg/platform/runtime/runtime.go

```diff
@@ -63,11 +75,33 @@ func newRuntime(target setup.Targeter, an analytics.Dispatcher, svcModel *model.
 		}
 	}
 
+	if target.ProjectDir() == "" || auth == nil {
+		return rt, nil
+	}
+
+	// Check if local build script has changes that should be staged.
+	if script, err := buildscript.NewScriptFromProjectDir(target.ProjectDir()); err == nil {
+		bp := model.NewBuildPlannerModel(auth)
+		if commitID, err := localcommit.Get(target.ProjectDir()); err == nil {
+			if expr, err := bp.GetBuildExpression(target.Owner(), target.Name(), commitID.String()); err == nil {
+				if !script.Equals(expr) {
+					return rt, &NeedsStageError{errs.New("Runtime changes should be staged")}
+				}
+			} else {
+				multilog.Error("Unable to get remote build expression: %v", errs.JoinMessage(err))
+			}
+		} else if !localcommit.IsFileDoesNotExistError(err) {
+			multilog.Error("Unable to get local commit: %v", errs.JoinMessage(err))
+		}
+	} else if !buildscript.IsDoesNotExistError(err) {
+		multilog.Error("Unable to get local build script: %v", errs.JoinMessage(err))
+	}
```

Comment:
I realized while reading this that we didn't cover for the offline use-case. Filed a followup here:

https://activestatef.atlassian.net/browse/DX-1934

It may be worth just doing that as this PR though, since much of the code being introduced here would effectively be partially throwaway. Up to you.